Cache basic authentications

roidelapluie opened the pull request:
In order to improve the response time with basic authentication, cache
the results of bcrypt in memory.

This is a draft to see if that approach would be acceptable.

The code is coming from Caddy, and I have fullfiled the attribution
required by Apache-2 by adding the copyright to our headers:
https://github.com/caddyserver/caddy/commit/9a7756c6e4b4ddb945bede3ddb2dfbf241208915

Signed-off-by: Julien Pivotto <roidelapluie@inuits.eu>

roidelapluie commented:
If the general approach is file I will add specific tests.

roidelapluie commented on web/users.go:
```diff
@@ -58,7 +62,21 @@ func (u *userAuthRoundtrip) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	if auth {
 		if hashedPassword, ok := c.Users[user]; ok {
-			if err := bcrypt.CompareHashAndPassword([]byte(hashedPassword), []byte(pass)); err == nil {
+			cacheKey := hex.EncodeToString(append(append([]byte(user), []byte(hashedPassword)...), []byte(pass)...))
```
We could also sha512 this one maybe if we don't want plaintext in memory.

SuperQ replied:
Basic auth is plain text anyway, so I'm not that concerned about it.

SuperQ commented on web/cache.go:
```diff
@@ -0,0 +1,77 @@
+package web
+
+import (
+	weakrand "math/rand"
```
Odd choice to go with this over `crypto/rand`.

roidelapluie replied:
It comes from caddy and I guess that we do not need more, it is just used for eviction, not for anything security related.

SuperQ replied:
Yea, I understand, I'm just interested in the choice.

SuperQ commented on web/cache.go:
```diff
@@ -0,0 +1,77 @@
+// Copyright 2020 The Prometheus Authors
```
```suggestion
// Copyright 2021 The Prometheus Authors
```

roidelapluie replied:
fixed

roidelapluie commented:
is this good to go?

roidelapluie commented:
Hold on.

roidelapluie commented:
OK now it works as expected. I have added getters and setters to release locks early.

I have also added a Lock around bcrypt so that we only run one at the same time.

SuperQ reviewed (approved):
LGTM

beorn7 commented on web/users.go:
```diff
@@ -40,6 +44,10 @@ type userAuthRoundtrip struct {
 	logger        log.Logger
+	cache         *cache
+	// bcryptMtx is there to ensure that bcrypt.CompareHashAndPassword is run
+	// only once in parallel as this is CPU expansive.
```
expansive → intensive?

roidelapluie replied:
🤦🏻 I will fix it
